pilot/pkg/xds/eds.go

```diff
@@ -161,13 +183,39 @@ func (s *DiscoveryServer) edsCacheUpdate(shard model.ShardKey, hostname string,
 		Namespace: namespace,
 	}: {}})
 	ep.mutex.Unlock()
+	if features.SendUnhealthyEndpoints && pushType == IncrementalPush {
+		// Check if new Endpoints are ready to be pushed. This check
+		// will ensure that if a new pod comes with a non ready endpoint,
+		// we do not unnecessarily push that config to Envoy.
+		// Please note that address is not a unique key. So this may not accurately
+		// identify based on health status and push too many times - which is ok since its an optimization.
+		emap := make(map[string]*model.IstioEndpoint, len(oldIstioEndpoints))
+		for _, oie := range oldIstioEndpoints {
+			emap[oie.Address] = oie
+		}
+		needPush := false
+		for _, nie := range istioEndpoints {
+			if oie, exists := emap[nie.Address]; exists {
+				// If endpoint exists already, we should push if it's health status changes.
+				needPush = oie.HealthStatus != nie.HealthStatus
```

Comment:
IIUC the original logic will do incremental push if there is any endpoint update event, and not limit to health status change. Now, we will only increment push if health status changes or there is new healthy endpoint. I am not sure if there is other part of EDS resource that could be changed due to endpoint update event?

Reply:
That did not change. It is same.